File: src/qat/common/ConsoleServer.java
package qat.common;

import java.net.*;
import java.io.*;
import java.util.*;

import qat.common.ProtocolConstants;

public class ConsoleServer extends Thread {
  
	private DataOutputStream clientOutput;
	private DataInputStream clientInput;
	private Socket clientSocket;
	private int clientDebugLevel;
  
	private static Vector consoleClientList;
	private static int debugLevel; // this is the level of detail we will use for debug message. 0 = none, 10 = max
  
	static {
		consoleClientList = new Vector();
	}
  
	private ConsoleServer(DataInputStream in, DataOutputStream out, Socket socket, int debugLevel) {
		clientInput = in;
		clientOutput = out;
		clientSocket = socket;
		clientDebugLevel = debugLevel;
		// if the client doesn't respond in 5 seconds, we assume it's disconnected
		// and terminate it's connection
		try {
			socket.setSoTimeout(5000);
		}
		catch (Exception e) {
			e.printStackTrace();
		}
	}
  
	private synchronized void clientDebugMsg(String msg, int level) throws Exception {
		if (level<=clientDebugLevel) {
			// write the output to this client
			clientOutput.writeUTF(msg+"\n");
			clientOutput.flush();
	  
			// read a byte to verify client is still active
			int code = clientInput.readInt();
			if (code != ProtocolConstants.RESPONSE_FINISHED_OK) {
				throw new Exception("Client not responding");
			}	
		}
	}
  
	public synchronized static void debugMsg(String msg) {
		debugMsg(msg,getDebugLevel());
	}
	
	public synchronized static void debugMsg(String msg,int level) {
		if (level<=getDebugLevel()) {
			System.out.println(msg);
		}
		ConsoleServer consoleClient;
		for (int i = 0; i < consoleClientList.size(); i++) {
			try {
				consoleClient = (ConsoleServer)consoleClientList.get(i);
				consoleClient.clientDebugMsg(msg,level);
			}
			catch (Throwable e) {
				if (getDebugLevel()>3) {
					System.out.println("Console client seems to have died :"+e.toString());
				}
				removeClient(i);
			}
		}
	}
	
	public synchronized static void debugStackTrace(Throwable ex) {
		StringWriter stringWriter = new StringWriter();
		ex.printStackTrace(new PrintWriter(stringWriter));
		debugMsg(stringWriter.toString());
		try {
			stringWriter.close();
		}
		catch (IOException err) {
		}
	}
  
	public synchronized static void setDebugLevel(int l) {
		debugLevel = l;
	}
  
	public synchronized static int getDebugLevel() {
		return debugLevel;
	}
  
	public synchronized static void removeClient(int i) {
		ConsoleServer consoleServer = (ConsoleServer)consoleClientList.get(i);	
		try {	  
			consoleServer.clientOutput.close();
		}
		catch (Throwable e) {
			e.printStackTrace();
		}
		try {	  
			consoleServer.clientInput.close();
		}
		catch (Throwable e) {
			e.printStackTrace();
		}
		try {	  
			consoleServer.clientSocket.close();
		}
		catch (Throwable e) {
			e.printStackTrace();
		}
		consoleClientList.remove(i);
	}
  
	public synchronized static void addClient(DataInputStream in, DataOutputStream out, Socket socket, int clientDebugLevel) {
		try {
			socket.setSoTimeout(1000);
		}
		catch (java.net.SocketException e) {
			e.printStackTrace();
		}
		consoleClientList.add(new ConsoleServer(in,out,socket,clientDebugLevel));
		debugMsg("There are now "+consoleClientList.size()+" client(s) registered on the console",3);
	}
  
	public synchronized static PrintStream getConsoleOutput() {
		return System.out;
	}
}

File: src/chorus/ConsoleServer.cpp
#include "ConsoleServer.hpp"

// define static variables here
int ConsoleServer::debugLevel = 10;// this is the level of detail we will use for debug message. 0 = none, 10 = max
ConsoleServer *ConsoleServer::consoleClientList[25];
int ConsoleServer::consoleClientListCount=0;

ConsoleServer::ConsoleServer(int socket, int thisdebugLevel) {
	clientSocket = socket;
	clientDebugLevel = thisdebugLevel;
}

ConsoleServer::~ConsoleServer()
{
}

int ConsoleServer::clientDebugMsg(char *msg, int level) {
	
  if (level<=clientDebugLevel) {

	// write the output to this client
	try {			
	  string msgString = string(msg);		
	  Utils::writeString(clientSocket,msgString);
	}
	catch (char *msg) {
	  return -1;			
	}
		
	try {			
	  // read a byte to verify client is still active
	  int code = Utils::readInt(clientSocket);
			
	  if (code != RESPONSE_FINISHED_OK) {
		return -1;			
	  }
	}
	catch (char *msg) {
	  return -1;			
	}			
  }
  return 0;	
}

void ConsoleServer::debugMsg(int level, const char *fmt_str, ...) 
{
	ConsoleServer *consoleClient;
	// retrieve the paramters
	char *str_cpy= new char[STR_LENGTH];		
	va_list args;
	va_start(args, fmt_str);			
	vsprintf(str_cpy, fmt_str, args);		
	va_end(args);
	
	if ((level <=  ConsoleServer::debugLevel)&&
		(stdout!=NULL)) {
		fprintf(stdout,str_cpy);
		fflush(stdout);			
	}
	
	for (int i = 0; i < getClientListCount(); i++) {
		try {
			consoleClient = (ConsoleServer*)consoleClientList[i];
			if (consoleClient->clientDebugMsg(str_cpy,level)!=0) {
				printf("Console client seems to have died\n");
				removeClient(i);
				printf("Removed client for Console server list\n");
			}
		}
		catch (char *ex) {
			printf("%s\n",ex);			
		}
	}	
	delete [] str_cpy;	
}

void ConsoleServer::debugMsgF(int level, FILE *fileHandle, const char *fmt_str, ...)
{	
	// retrieve the paramters
	char *str_cpy= new char[STR_LENGTH];		
	va_list args;
	va_start(args, fmt_str);			
	vsprintf(str_cpy, fmt_str, args);		
	va_end(args);
		
	// now write to stdout if required
	if ((level<= ConsoleServer::debugLevel)&&
		(fileHandle!=NULL)) {		
	
		// to console
		if (fileHandle != stdout) {
			printf(str_cpy);
		}
		
		// to file now
		if (fileHandle>0) {
			fprintf(fileHandle,str_cpy);
			fflush(fileHandle);		
			va_end(args);
		}		
		
		delete [] str_cpy;		
	}
}

void ConsoleServer::setDebugLevel(int l) 
{
	ConsoleServer::debugLevel = l;
}
  
int ConsoleServer::getDebugLevel() {
	return ConsoleServer::debugLevel;
}

int ConsoleServer::getClientListCount() {
  return consoleClientListCount;  
}

void ConsoleServer::removeClient(int i) {
	
	if (consoleClientListCount==0)
		return;
	
	ConsoleServer *consoleServer = (ConsoleServer *)consoleClientList[i];	
 	try {	  
		close(consoleServer->clientSocket);
		delete consoleServer;		
 	}
 	catch (char * e) {
		printf("%s\n",e);
 	}
	for (i = i; i < getClientListCount(); i++)
		consoleClientList[i]=consoleClientList[i+1];
	consoleClientList[getClientListCount()]=NULL;
	consoleClientListCount--;
}
  
void ConsoleServer::addClient(int socket, int thisclientDebugLevel) {
	consoleClientList[getClientListCount()] = new ConsoleServer(socket,thisclientDebugLevel);
	consoleClientListCount++;
	
	debugMsg(3,"There are now %d client(s) registered on the console\n",getClientListCount());
}
